### PR TITLE
[WIP]Add support to run tests agains Python 3 using tox and multiprocessing support

### PR DESCRIPTION
**Pull request by wiliamsouza**
My plan was to open two PR one to add Python 3 test support and another with multiprocessing support
but it seems not possible.

This adds:

* `python setup.py test` command.
* `tox` command.

Examples:

```python
>>> import atomos.multiprocessing.atomic
>>> counter = atomos.multiprocessing.atomic.AtomicInteger()
>>> counter.get()
0
```

and 

```python
from multiprocessing import Process

process_count=10
loop_count=1000

atomic_int = atomos.multiprocessing.atomic.AtomicInteger()

def inc_atomic_int():
    for _ in range(loop_count):
        atomic_int.add_and_get(1)

processes = []
for _ in range(process_count):
    p = Process(target=inc_atomic_int)
    processes.append(p)
    p.start()
```

### CODE REVIEW

**Comment by maxcountryman** [comment]
Hi William, thanks for this! Looks promising. I'll have to have a closer look a little later on today.

File: atomos/atomic.py [hunk 1]
```diff
@@ -157,7 +162,7 @@ def __init__(self, value=0):
 
     def __setattr__(self, name, value):
         # Ensure the `_value` attribute is always an int.
-        if name == '_value' and not isinstance(value, types.IntType):
+        if name == '_value' and not isinstance(value, int):
```
Comment (maxcountryman) [inline atomos/atomic.py]: Should this still be types.IntType for Python 2?

File: atomos/atomic.py [hunk 2]
```diff
@@ -172,7 +177,7 @@ def __init__(self, value=long(0)):
 
     def __setattr__(self, name, value):
         # Ensure the `_value` attribute is always a long.
-        if name == '_value' and not isinstance(value, types.LongType):
+        if name == '_value' and not isinstance(value, long):
```
Comment (maxcountryman) [inline atomos/atomic.py]: Ditto.

File: atomos/atomic.py [hunk 3]
```diff
@@ -187,7 +192,7 @@ def __init__(self, value=float(0)):
 
     def __setattr__(self, name, value):
         # Ensure the `_value` attribute is always a float.
-        if name == '_value' and not isinstance(value, types.FloatType):
+        if name == '_value' and not isinstance(value, float):
```
Comment (maxcountryman) [inline atomos/atomic.py]: Ditto.

File: atomos/multiprocessing/atomic.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+import six
+
+import atomos.util as util
+
+if six.PY3:
+    long = int
+
+
+# TODO: Fix it to work with other types
```
Comment (maxcountryman) [inline atomos/multiprocessing/atomic.py]: What are the current limitations?

File: atomos/multiprocessing/atomic.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+# https://docs.python.org/3.4/library/ctypes.html#fundamental-data-types
+class AtomicReference(object):
+    '''
+    A reference to an object which allows atomic manipulation semantics.
+
+    AtomicReferences are particularlly useful when an object cannot otherwise
+    be manipulated atomically.
+    '''
+    def __init__(self, typecode=None, value=None):
```
Comment (maxcountryman) [inline atomos/multiprocessing/atomic.py]: What's the `typecode` parameter?

Comment (maxcountryman) [inline atomos/multiprocessing/atomic.py]: Ah, never mind, I see it used below--still probably worth a docstring.

File: tests/test_multiprocessing_atomic.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+from multiprocessing import Process
+
+import pytest
+
+import atomos.multiprocessing.atomic
+
+
+"""
+@pytest.fixture
```
Comment (maxcountryman) [inline tests/test_multiprocessing_atomic.py]: WIP?

**Comment by maxcountryman** [comment]
Overall it looks really good! I have a couple of comments on the commit set. I think we should also probably update the README section regarding multiprocessing to explain how support for it has evolved.

File: atomos/atomic.py (hunk 2 above)
Comment (wiliamsouza) [inline atomos/atomic.py]: Python 3 don't have this type any more. See http://www.diveintopython3.net/porting-code-to-python-3-with-2to3.html#long

Not sure but using the primitive type can do the job for both Python versions.

File: atomos/multiprocessing/atomic.py (hunk 4 above)
Comment (wiliamsouza) [inline atomos/multiprocessing/atomic.py]: Python 3 int() and long() unification as described above.

File: atomos/multiprocessing/atomic.py (hunk 5 above)
Comment (wiliamsouza) [inline atomos/multiprocessing/atomic.py]: I will provide a docstring.

File: atomos/atomic.py (hunk 2 above)
Comment (maxcountryman) [inline atomos/atomic.py]: Okay makes sense. In the case that we're running this code in Python 2 do you think we should still use `types`?

File: tests/test_multiprocessing_atomic.py (hunk 6 above)
Comment (wiliamsouza) [inline tests/test_multiprocessing_atomic.py]: AtomicReference only accept ctypes[2] objects supported by `multiprocessing.Value`[1]

[1] https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Value
[2] https://docs.python.org/3/library/ctypes.html#fundamental-data-types

File: atomos/atomic.py (hunk 2 above)
Comment (wiliamsouza) [inline atomos/atomic.py]: We can swap `types.LongType` to `long` and keep the other types using `types.*`.

Comment (maxcountryman) [inline atomos/atomic.py]: Yeah that seems like it'd make sense.

File: tests/test_multiprocessing_atomic.py (hunk 6 above)
Comment (maxcountryman) [inline tests/test_multiprocessing_atomic.py]: Okay so this means no dictionaries, for example? Seems like a pretty important limitation. I think we should update the README in this patch.

**Comment by wiliamsouza** [comment]
@maxcountryman I made some changes as PR comments. But I'm blocked with a ReaderWriterLock test. Please take a look.

**Comment by maxcountryman** [comment]
Sure. I should be able to look this weekend.

**Comment by maxcountryman** [comment]
@wiliamsouza not sure about the `ReaderWriterLock` issue...haven't had success tracking down the issue so far.

**Comment by maxcountryman** [comment]
I think the issue may be related to the fact that incrementing integers between processes is not atomic. Also I'm not sure if those counters are even shared. I'm guessing that using a `multiprocessing.Value` for `_reader_count` is necessary but so far I'm not able to get that to work.

**Comment by wiliamsouza** [comment]
@maxcountryman thanks, will take a look!

File: atomos/atomic.py (hunk 1 above)
Comment (wiliamsouza) [inline atomos/atomic.py]: `IntType` == `int` in Python3

**Comment by wiliamsouza** [comment]
`AtomicReference` works like the thread version!

Just removing the Multiprocessing note is enough? What do you think?

@maxcountryman ^

**Comment by maxcountryman** [comment]
@wiliamsouza sure, removing that is probably enough. I guess I thought now you could only store Types with it though?

**Comment by wiliamsouza** [comment]
Look this [1] and this [2].
Now we have two `AtomicReference` implementation one of it only support `ctypes` and the other  works like the threaded version. I think it removes the restrictions we talked later.

[1] https://github.com/maxcountryman/atomos/pull/2/files#diff-a5fa8912dea8db2cd0aa83511c1f55baR21
[2] https://github.com/maxcountryman/atomos/pull/2/files#diff-a5fa8912dea8db2cd0aa83511c1f55baR33

**Comment by maxcountryman** [comment]
Awesome! All tests passing now too?

**Comment by wiliamsouza** [comment]
like a charm :D 
```
  py27: commands succeeded
  py32: commands succeeded
  py33: commands succeeded
  py34: commands succeeded
  pypy: commands succeeded
  congratulations :)
```

**Comment by maxcountryman** [comment]
Nice work!

**Comment by maxcountryman** [comment]
I think I may have found a bug in `atomos.multiprocessing.AtomicReference`, see #3.

Here's a simple test case:

```diff
diff --git a/tests/test_multiprocessing_atomic.py b/tests/test_multiprocessing_atomic.py
index f8600b4..9548b2c 100644
--- a/tests/test_multiprocessing_atomic.py
+++ b/tests/test_multiprocessing_atomic.py
@@ -79,3 +79,26 @@ def test_concurrent_atomic_int(process_count=10, loop_count=1000):
         p.join()
 
     assert atomic_int.get() == process_count * loop_count
+
+
+def test_concurrent_atomic_ref(process_count=10, loop_count=1000):
+    atomic_ref = atomos.multiprocessing.atomic.AtomicReference(0)
+
+    def inc_atomic_ref():
+        for _ in range(loop_count):
+            while True:
+                oldval = atomic_ref.get()
+                is_set = atomic_ref.compare_and_set(oldval, oldval + 1)
+                if is_set:
+                    break
+
+    processes = []
+    for _ in range(process_count):
+        p = Process(target=inc_atomic_ref)
+        processes.append(p)
+        p.start()
+
+    for p in processes:
+        p.join()
+
+    assert atomic_ref.get() == process_count * loop_count
```

The `atomic_ref` ends up returning `0` so it seems that memory between processes is not being shared as we'd expect.

**Comment by wiliamsouza** [comment]
@maxcountryman Can you move the text above as comment of #3?

**Pull request merged**